imports/methods/TypedMethod.ts

```diff
@@ -0,0 +1,67 @@
+import { check } from 'meteor/check';
+import { EJSONable, EJSONableProperty } from 'meteor/ejson';
+import { Meteor } from 'meteor/meteor';
+
+type TypedMethodParam = EJSONable | EJSONableProperty
+type TypedMethodArgs = Record<string, TypedMethodParam>;
+type TypedMethodValidator<Arg extends TypedMethodArgs> =
+  (this: Meteor.MethodThisType, arg0: Record<keyof Arg, unknown>) =>
+    Arg;
+type TypedMethodRun<Arg extends TypedMethodArgs, Return extends TypedMethodParam | void> =
+  (this: Meteor.MethodThisType, arg0: Arg) => Return;
+
+export default class TypedMethod<
+  Args extends TypedMethodArgs,
+  Return extends TypedMethodParam | void,
+> {
+  name: string;
+
+  validate: TypedMethodValidator<Args>;
+
+  run: TypedMethodRun<Args, Return>;
+
+  constructor({ name, validate, run }: {
+    name: string,
+    validate: TypedMethodValidator<Args>,
+    run: TypedMethodRun<Args, Return>,
+  }) {
+    check(name, String);
+    check(validate, Function);
+    check(run, Function);
+
+    this.name = name;
+    this.validate = validate;
+    this.run = run;
+
+    /* eslint-disable-next-line @typescript-eslint/no-this-alias -- we need to
+       capture both the TypedMethod object (so we can find the implementation)
+       and the method invocation context (for things like userId) */
+    const self = this;
+    Meteor.methods({
+      /* eslint-disable-next-line meteor/audit-argument-checks -- because we're
+         wrapping the method call, the eslint check can't tell what's going on,
+         but the actual Meteor audit-argument-checks package will still fire if
+         we don't use check methods. */
+      [this.name](arg0: Args) {
```

Comment:
You can potentially punt on solving this until you get to the first method implementation that does a server-only import or looks at a server-only collection, but I suspect eventually we're going to get to a situation where there's an import chain:

* client/main -> 
* something in imports/client/components/* (which wants to make a `call`) -> 
* something in imports/methods/* ->
* something in imports/server/* (where `run` accesses some server-only collection or imports some server-only code)

and then the build tool will be like "you can't import a thing under server from client" and there might not be a good place to do a conditional import.

`reorganizeGoogleDrive` and `setupGetUploadToken` are both examples of methods that will require server-only imports, but which will have the client code importing the file under `methods/` -- maybe try porting one of those as a pressure test?

In turn, that may prompt us to think about if we want to always be defining these methods on both client-side and server-side, or if we'd be better off keeping the bundle smaller and not shipping method implementations to the client that are not useful to the user experience.

Reply:
Pushed a new version that does this, with a strawman structure.